wasm/main.go

```diff
@@ -73,3 +74,30 @@ func makeDid(_ js.Value, args []js.Value) interface{} {
 	return js.ValueOf(resultObj)
 
 }
+
+func resolveDid(_ js.Value, args []js.Value) interface{} {
```

Comment:
do we have a way to distinguish in the JS if this was an error vs the return object?

Reply:
that's a great question. I've had a similar question and I've not been able to determine a way aside from serializing a string with the word "error". Maybe the right way is to build a wrapper around the objects such as the following:

```go
type ObjectType string

TypeError ObjectType = "error"
TypeObject ObjectType = "object"

type ObjectWrapper  {
     ObjectType ObjectType `json:"type"`
     Object interface{} `json:"object"`
}
```

or something. There's no "standard" way though, AFAIK.

Reply:
@nitro-neal  yeah there's no easy way to throw a JS exception from web assembly to support something like

```js
try {
  resolveDid("did:janky:alice")
} catch (e) {
  console.log(error);
}
```

WASM execution happens within an isolated sandbox in a separate execution environment that cannot directly access the JS stack. So anything thrown from WASM gets caught by the WASM runtime and handled as if the WASM code panicked.

@michaelneale your top level `error` property could definitely work. Another option that may feel a bit more like catching an exception would be to return a `Promise` from WASM that is `resolve`d or `reject`ed. Using `await` semantics you'd be able to try/catch the returned promise, e.g.

```js
try {
  await resolveDid("did:janky:alice")
} catch (e) {
  console.log(error);
}
```

the underlying Go code would look like this:
```go
func resolveDid(_ js.Value, args []js.Value) interface{} {
    handler := js.FuncOf(func(this js.Value, args []js.Value) interface{} {
        resolve := args[0]
        reject := args[1]
    
        go func() {
            didString := args[0].String()
            resolvers := []did.Resolution{did.KeyResolver{}, did.WebResolver{}, did.PKHResolver{}, did.PeerResolver{}}
            resolver, err := did.NewResolver(resolvers...)
            
            if err != nil {
                // err should be an instance of `error`, eg `errors.New("some error")`
                errorConstructor := js.Global().Get("Error")
                errorObject := errorConstructor.New(err.Error())
                reject.Invoke(errorObject)
            }
    
            doc, err := resolver.Resolve(didString)
            if err != nil {
                errorConstructor := js.Global().Get("Error")
                errorObject := errorConstructor.New(err.Error())
                reject.Invoke(errorObject)
            }
    
            resultBytes, err := json.Marshal(doc)
            if err != nil {
                errorConstructor := js.Global().Get("Error")
                errorObject := errorConstructor.New(err.Error())
                reject.Invoke(errorObject)
            }
    
            var resultObj map[string]interface{}
            err = json.Unmarshal(resultBytes, &resultObj)
            if err != nil {
                errorConstructor := js.Global().Get("Error")
                errorObject := errorConstructor.New(err.Error())
                reject.Invoke(errorObject)
            }
    
            resolve.Invoke(js.ValueOf(resultObj))
        }()
    
        return nil
    })
    
    promiseConstructor := js.Global().Get("Promise")
    return promiseConstructor.New(handler)
}
```

the go is def a bit uglier but i bet we can teach ChatGPT to return functions with this kind of wrapper.


random unnecessary info:
the "uneasy" way to throw exceptions from WASM into the JS runtime i think would require us to provide the assembler with a hint. so something like:

```go
// throw stub in our go code
func Throw(exception string, message string)
```

```
// assembler hint in ssi-sdk_js.s

TEXT ·Throw(SB), NOSPLIT, $0
  CallImport
  RET
```

```js
// manually added to wasm_exec.js into `importObject.go` object
// this object already exists in `wasm_exec.js`, just adding here for sake of example
this.importObject = {
// `go` already exists in `wasm_exec.js`. just adding here for sake of example
  go: {
    // this is what we would need to add ourselves
    // func Throw(exception string, message string)
    'ssi-sdk.Throw': (sp) => {
      const exception = loadString(sp + 8)
      const message = loadString(sp + 24)
      const throwable = globalThis[exception](message)
      throw throwable
    }
  }
}
```

Reply:
Heavy +1 to returning a promise. DID resolution will typically involve making a network request, which reinforces the fact that it should be an awaitable. 

In fact, I can see the [Resolve](https://github.com/TBD54566975/ssi-sdk/blob/92be548d54be11b3740b95bb5c568958e5979de0/did/resolver.go#L16) method evolving to become more go idiomatic and accepting a `context.Context` object as the first parameter. This object would carry cancellation signals.